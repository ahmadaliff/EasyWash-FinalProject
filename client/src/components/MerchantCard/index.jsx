import PropTypes from 'prop-types';
import _ from 'lodash';
import { useState } from 'react';
import toast from 'react-hot-toast';
import { FormattedMessage } from 'react-intl';
import { useNavigate } from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { Avatar, Card, CardMedia, IconButton, Skeleton } from '@mui/material';
import { Favorite, LocalShipping } from '@mui/icons-material';

import config from '@config/index';

import intlHelper from '@utils/intlHelper';

import { selectLogin, selectUser } from '@containers/Client/selectors';
import { actionAddToFavorit, actionDeleteFromFavorit } from '@pages/Favorit/action';

import classes from '@components/MerchantCard/style.module.scss';

const MerchantCard = ({ user, login, merchant }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [loading, setLoading] = useState(true);
  const handleNavigate = () => {
    if (!login) navigate(`/login`);
    else if (user?.role === 'user') navigate(`/user/laundry/${merchant?.id}`);
    else toast.error(intlHelper({ message: 'app_must_user_role' }));
  };
  return (
    <div className={classes.cardWrap} data-testid="card-wrap">
      <Card className={classes.card} onClick={handleNavigate} data-testid="card-merchant">
        {merchant?.imagePath ? (
          <>
            {loading && <Skeleton variant="rectangular" height="194" width="100%" />}
            <CardMedia
              component="img"
              className={classes.img}
              image={`${config.api.server}${merchant?.imagePath}`}
              alt={merchant?.name}
              loading="lazy"
              onLoad={() => setLoading(false)}
            />
          </>
        ) : (
          <Avatar className={classes.avatar} variant="square">
            {merchant?.name?.split(' ')[0][0]}
            {merchant?.name?.split(' ').length > 1 && merchant?.name?.split(' ')[1][0]}
          </Avatar>
        )}
        <div className={classes.logoDelivery}>
          <LocalShipping />
          <FormattedMessage id="app_free_delivery" />
        </div>
        <h4 className={classes.headerText}>{merchant?.name}</h4>
        <p className={classes.distance}>
          {merchant?.distance} {merchant?.distance !== undefined && 'Km'}
        </p>
        <p className={classes.desc}>{merchant?.description}</p>
      </Card>
      {login && user?.role === 'user' && (
        <div className={classes.cardAction}>
          {!_.find(merchant.Favorits, (val) => val?.userId === user?.id) ? (
            <IconButton onClick={() => dispatch(actionAddToFavorit(merchant?.id))} data-testid="favorit-button">
              <Favorite />
            </IconButton>
          ) : (
            <IconButton onClick={() => dispatch(actionDeleteFromFavorit(merchant?.id))}>
              <Favorite className={classes.favoritRed} data-testid="favorit-red-button" />
            </IconButton>
          )}
        </div>
      )}
    </div>
  );
};

MerchantCard.propTypes = {
  merchant: PropTypes.object,
  user: PropTypes.object,
  login: PropTypes.bool,
};

const mapStateToProps = createStructuredSelector({
  user: selectUser,
  login: selectLogin,
});

export default connect(mapStateToProps)(MerchantCard);
