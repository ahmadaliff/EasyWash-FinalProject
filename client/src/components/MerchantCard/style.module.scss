@import '@styles/base';
.cardWrap {
  position: relative;
  .card {
    width: 15rem;
    box-shadow: 0 0 0.25rem gray;
  }
  .headerText {
    margin-top: 0;
  }
  .desc {
    font-size: small;
    display: -webkit-box;
    -webkit-line-clamp: 4;
    -webkit-box-orient: vertical;
    overflow: hidden;
  }
  .cardAction {
    position: absolute;
    top: 0;
    right: 0;
  }
  .avatar {
    height: 194px;
    width: 100%;
    background-color: var(--color-primary);
  }
}
