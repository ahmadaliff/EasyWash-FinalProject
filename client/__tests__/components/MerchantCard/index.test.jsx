import MerchantCard from '@components/MerchantCard';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;
const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));
jest.mock('@containers/Client/selectors', () => ({
  selectLogin: () => true,
  selectUser: () => ({
    id: 1,
    fullName: 'ahmad alif sofian',
    role: 'user',
    imagePath: null,
  }),
}));

const mockProps = {
  merchant: {
    id: 123,
    name: 'merchant',
    imagePath: null,
    distance: 5,
    Favorits: [],
  },
};

beforeEach(() => {
  wrapper = render(<MerchantCard {...mockProps} />);
});
describe('MerchantCard Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('card-merchant')).toBeInTheDocument();
    expect(getByTestId('card-wrap')).toBeInTheDocument();
  });
  test('Should be Navigate onclick card', () => {
    const { getByTestId } = wrapper;
    fireEvent.click(getByTestId('card-merchant'));
    expect(mockNavigate).toHaveBeenCalledWith(`/user/laundry/123`);
  });
  test('Dispatches actionAddToFavorit when the merchant is not favorited', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('favorit-button')).toBeInTheDocument();
    fireEvent.click(getByTestId('favorit-button'));
    expect(mockDispatch).toBeCalled();
  });
  test('Dispatches actionDeleteFromFavorit when the merchant is favorited', () => {
    const favoritedProps = {
      ...mockProps,
      merchant: {
        ...mockProps.merchant,
        Favorits: [{ userId: 1, merchantId: 123 }],
      },
    };
    wrapper.rerender(<MerchantCard {...favoritedProps} />);
    const { getByTestId } = wrapper;
    expect(getByTestId('favorit-red-button')).toBeInTheDocument();
    fireEvent.click(getByTestId('favorit-red-button'));
    expect(mockDispatch).toBeCalled();
  });

  test('Should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
