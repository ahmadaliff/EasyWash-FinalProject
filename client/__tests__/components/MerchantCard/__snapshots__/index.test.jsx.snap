// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantCard Component Should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cardWrap"
        data-testid="card-wrap"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card css-bhp9pd-MuiPaper-root-MuiCard-root"
          data-testid="card-merchant"
        >
          <div
            class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
          >
            m
          </div>
          <div
            class="logoDelivery"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LocalShippingIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 8h-3V4H3c-1.1 0-2 .9-2 2v11h2c0 1.66 1.34 3 3 3s3-1.34 3-3h6c0 1.66 1.34 3 3 3s3-1.34 3-3h2v-5l-3-4zM6 18.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm13.5-9 1.96 2.5H17V9.5h2.5zm-1.5 9c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
              />
            </svg>
            Gratis Ongkir
          </div>
          <h4
            class="headerText"
          >
            merchant
          </h4>
          <p
            class="distance"
          >
            5
             Km
          </p>
          <p
            class="desc"
          />
        </div>
        <div
          class="cardAction"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="favorit-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cardWrap"
      data-testid="card-wrap"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root card css-bhp9pd-MuiPaper-root-MuiCard-root"
        data-testid="card-merchant"
      >
        <div
          class="MuiAvatar-root MuiAvatar-square MuiAvatar-colorDefault avatar css-18gxrea-MuiAvatar-root"
        >
          m
        </div>
        <div
          class="logoDelivery"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="LocalShippingIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20 8h-3V4H3c-1.1 0-2 .9-2 2v11h2c0 1.66 1.34 3 3 3s3-1.34 3-3h6c0 1.66 1.34 3 3 3s3-1.34 3-3h2v-5l-3-4zM6 18.5c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm13.5-9 1.96 2.5H17V9.5h2.5zm-1.5 9c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5z"
            />
          </svg>
          Gratis Ongkir
        </div>
        <h4
          class="headerText"
        >
          merchant
        </h4>
        <p
          class="distance"
        >
          5
           Km
        </p>
        <p
          class="desc"
        />
      </div>
      <div
        class="cardAction"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="favorit-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
